src/marks/density.js

```diff
@@ -12,43 +12,117 @@ const defaults = {
 
 export class Density extends Mark {
   constructor(data, options = {}) {
-    const {x, y, bandwidth = 20, thresholds = 20} = options;
+    const {x, y, z, weight, bandwidth = 20, thresholds = 20} = options;
     super(
       data,
       [
         {name: "x", value: x, scale: "x", optional: true},
-        {name: "y", value: y, scale: "y", optional: true}
+        {name: "y", value: y, scale: "y", optional: true},
+        {name: "weight", value: weight, optional: true},
+        {name: "z", value: maybeZ(options), optional: true}
       ],
       options,
       defaults
     );
     this.bandwidth = +bandwidth;
     this.thresholds = +thresholds;
+    this.z = z;
+    this.path = geoPath();
+  }
+  filter(index) {
+    return index;
   }
   render(index, {x, y}, channels, dimensions) {
-    const {x: X, y: Y} = channels;
-    const {dx, dy, bandwidth, thresholds} = this;
-    const [cx, cy] = applyFrameAnchor(this, dimensions);
-    const {width, height} = dimensions;
+    const {contours} = channels;
+    const {dx, dy, path} = this;
     return create("svg:g")
         .call(applyIndirectStyles, this, dimensions)
         .call(applyTransform, x, y, offset + dx, offset + dy)
-        .call(g => g.selectAll("path")
-          .data(contourDensity()
-              .x(X ? i => X[i] : constant(cx))
-              .y(Y ? i => Y[i] : constant(cy))
-              .size([width, height])
-              .bandwidth(bandwidth)
-              .thresholds(thresholds)
-            (index))
+        .call(g => g.selectAll()
+          .data(Array.from(index, i => [i]))
           .enter()
           .append("path")
-            .attr("d", geoPath()))
-      .node();
+            .call(applyDirectStyles, this)
+            .call(applyGroupedChannelStyles, this, channels)
+            .attr("d", ([i]) => path(contours[i])))
+        .node();
   }
 }
 
-export function density(data, {x, y, ...options} = {}) {
+export function density(data, {x, y, stroke, fill, bandwidth = 20, thresholds = 20, ...options} = {}) {
   ([x, y] = maybeTuple(x, y));
-  return new Density(data, {...options, x, y});
+  let f, s;
+  if (fill === "density") { fill = undefined; f = true; }
+  if (stroke === "density") { stroke = undefined; s = true; }
+  return new Density(data, {...options, x, y, fill, stroke, initializer: initializer(+bandwidth, +thresholds, f, s)});
```

Comment:
We should move this initializer to inside the Density class constructor, not the convenience constructor here. (You’d never use the Density class without it.)